### DISCUSSION TITLE
Resource for Learning

### DISCUSSION THREAD

**Discussion by mohammadazeemwani** [kerolloz/kerolloz#2] [Q&A] [answered]
In my projects my authentication part is messed up 
sometimes i don't know what i am doing 
like which technique to use where and unlike other libraries npm has these authentication ones kinda gets over my head 
have you any guide that will teach me like what is really happening
btw i use fastify these days for development

**Comment by kerolloz** [comment, marked as answer]
Hey @mohammadazeemwani,

Thank you for asking. I'll try to address your question. So let's start with the concepts:

## Authentication vs. Authorization:

- **Authentication**: Verifying who a user is (e.g., login using username/password).
- **Authorization**: Determining what a user is allowed to do (e.g., access control, permissions).

## Types of Authentication:

- **Session-based Authentication**: Stores user credentials in a session (cookie) after they log in. The server verifies these credentials on each request.
- **Token-based Authentication**: Involves issuing a token (like JWT) after login, which is passed along with every request for validation.
- **OAuth/OpenID Connect**: Delegates authentication to another provider (like Google, Facebook) and returns an access token that allows a user to be logged in.

## Fastify Authentication Libraries

### fastify-jwt:
- **What it does**: Adds JWT-based authentication to Fastify.
- **When to use**: Use this when you want to handle authentication statelessly with tokens (e.g., for APIs).
- **How it works**:
  - User logs in → Server sends back a JWT token.
  - On subsequent requests, the client sends the token in headers (typically the `Authorization` header).
  - The server verifies the token using a secret or public key.

### fastify-cookie + fastify-session:
- **What it does**: Implements session-based authentication.
- **When to use**: This is useful for traditional web apps where you maintain user sessions.
- **How it works**:
  - After login, the server sets a session ID in a cookie on the client.
  - For every request, the server checks the session store to see if the session ID is valid.

### fastify-oauth2:
- **What it does**: Handles OAuth 2.0-based authentication (e.g., using Google, GitHub).
- **When to use**: If you want to offload authentication to third-party providers.
- **How it works**:
  - Your app redirects users to an OAuth provider (Google, Facebook).
  - The user logs in, and the provider returns an access token.
  - Your server uses that token to authenticate the user.

## Choosing the Right Method

### Web Applications:
- Use **session-based authentication** if you’re building a traditional web app.
- Use **OAuth** if you want to allow login with third-party providers (like Google).

### APIs:
- Use **JWT authentication** for stateless, token-based authentication in REST APIs.

## General Flow (Using JWT as an Example)

1. **Login**:
   - User submits credentials → Server checks credentials.
   - If valid, server creates a JWT token and sends it to the client.
   
2. **Client-side**:
   - The client stores the token (in `localStorage` or cookies).
   - For every subsequent request, the token is sent in the headers.
   
3. **Validation**:
   - The server checks the token, verifies its signature, and if valid, allows the request.

---

I recommend starting with the **fastify-jwt** plugin if you're working with APIs. Here's an [official Fastify JWT guide](https://github.com/fastify/fastify-jwt) that might help you dive deeper. Let me know if you want more specific code examples!
